### ISSUE TITLE
Wrong detection of unsaved files

### ISSUE THREAD

**Issue by narma** [Microsoft/vscode#919]
Vscode shows me I have unsaved changes:
![image](https://cloud.githubusercontent.com/assets/91045/11530276/235e4580-9915-11e5-8e63-f6f2e39159b4.png)

Ok, before I saved it, I check md5 for this file.
![image](https://cloud.githubusercontent.com/assets/91045/11530264/09f00110-9915-11e5-87eb-2566c3bb26b4.png)

After save nothing changes.
![image](https://cloud.githubusercontent.com/assets/91045/11530290/3775d4a2-9915-11e5-8efc-283ae880cce0.png)

It happens time to time and it's very annoying.

**Comment by narma** [comment]
Btw, it's reproducible in my current session, steps:
1. Save
2. Click (focus) current file content.

Also, I checked this bug without extensions - bug not appears. It only happens with parinfer extension I use. Not sure it's parinfer or vscode bug, but think it's vscode.

Vscode version: 0.10.3
Parinfer ext: https://github.com/Microsoft/vscode-parinfer

**Comment by narma** [comment]
cc @joaomoreno

**Comment by joaomoreno** [comment]
The `parinfer` extension will always provide the buffer to the parinfer lib, which will return what the buffer should look like. It can be that parinfer returns exactly the same text as it was given. But the extension will simply make the text edit no matter what.

@bpasero should the extension not replace the text if it detects it hasn't changed, or should we finally implement content equals? You could be smart (good luck) and first compare the length and only then run it through a fast hashing function to compare contents and figure out whether the buffer has really changed.

**Comment by bpasero** [comment]
Ok this explains it. Nice idea, but not really a bug I would say.

**Comment by csholmq** [comment]
@joaomoreno @bpasero I can reproduce this without the Parinfer extension.

1. Open a file
2. Press space
3. Remove space

=> File still marked as dirty

1. Open a file
2. Press space
3. Press Ctrl+Z

=> File no longer dirty

**Comment by rightaway** [comment]
I thought when doing ctrl+z it still shows as dirty. Maybe this was changed recently.

**Comment by suan** [comment]
Yes, would be great if the dirty indicator only shows up for any actual changes - today it shows even after a change has been made and has been undone

**Comment by samuela** [comment]
Atom has this feature. If you add something to a file, change your mind and then remove those changes the file is marked as up-to-date. It's a simple feature but it's so incredibly intuitive and satisfying for the user experience side of things.

**Comment by samuela** [comment]
What would it take to implement this?

**Comment by shanemcgraw** [comment]
@bpasero any way this can be looked into? Would love to hear (to @samuela's comment) what it would take to get this shipped with Code. It's literally the only thing keeping me from making the switch from Atom 😄

**Comment by bpasero** [comment]
@alexandrudima @rebornix is there any kind of checksum you guys keep in the model based on the content that we could use to identify (or mark) the state as it is on disk and find out when certain edit operations managed to end up in that exact same state again?

My biggest issue with computing some checksum is that it is not clear to me what the performance impact of doing this is. I would stay away from computing this checksum each time the user makes modifications, so there must be a more clever way that has no performance penalty.

**Comment by alexandrudima** [comment]
The buffer has no such concept at this time...

**Comment by usernamehw** [comment]
@joaomoreno, is it possible to use `git` for this case and update editor's dirty state?

**Comment by GrCOTE7** [comment]
Hi.

Is that possible to unable this feature ? 
(The control of dirty file...)

**Comment by ericbf** [comment]
Any update on this issue?

**Comment by schontz** [comment]
Can you at least remove the dirty flag if all changes have been undone? If I open a file, make a change, hit undo, it should no longer be dirty.

All the other editors I've routinely used seem to understand that concept. Hashing for diffs is neat, but for large files it would be a performance hit. I would expect something like `if (undoStackAtSave == undoStackNow) notDirty()`. You get the idea.

**Comment by narma** [comment]
It's about user expierence. 
Dirty flag means something was changed. Not ideally, but acceptable if dirty flag setted when editor have history but no actually changes happenned, when user input has initiated it, for ex. after press space and then backspace). No changes actually here, but still we have a history. 
And user for sure not expected when dirty flag is set as a result on any plugin action (code reformat at this case) which in result doesn't apply any changes to editor.

**Comment by anru** [comment]
Hashing is not an option here. Hashing could be helpful if you really need some "snapshot" of the file and you don't have resources or other reasons to keep file entirely. But in case of local file editing you have full file contents for free.

So, you can do `isBuffersEquals(fileDiskContents, bufferContents)` instead of calculating hashes first (why ?) and compare them later. In case of buffers are not equal you can return early as first difference occurs, in case of calculating hashes – you can not. 

@schontz 's idea is nice too, if vscode internal architecture allows to do like it

**Comment by yaaase** [comment]
👍 to all previous requests, it would be nice to not see the dirty file indicator if the file state matches the previous written state.

**Comment by runfaj** [comment]
Nuts that this is still a problem after 4 years. It seems like even if it isn't the default, changing how this compares as an option is essential. It's such an integral part of many people's workflows.

**Comment by jackcasey** [comment]
This is annoying and kind of doing my head in. Little unsaved circles always on my editors. I think I use the 'fiddle a bit', 'undo it all' workflow a lot. Also using VIM `u` for undo plays even less nicely than `cmd-z`.

**Comment by sumagnadas** [comment]
i recently changed from Atom to VS Code. I didn't like this dirty file indicator even when there were no changes so can anyone please see into this. This is a great feature and as @runfaj says,

>It's such an integral part of many people's workflows.

So i would be glad if this feature comes to VS Code.

**Comment by alexschcom** [comment]
@bpasero I saw your comment on a duplicate issue and I get the argument that it's an indication that there's an undo-stack. But if you want to undo something, you just hit Ctrl-Z anyway. I think there's a lot more use in having it function as a way of telling if the current state of the file is equal to the one already saved on disk.

For example copy-pasting a chunk of code and immediately knowing that it's the same as what's already saved. Also, to avoid confusion about saving the exact same state of a file without knowing if there have actually been made changes.

Please at least provide a setting to change the behavior of the "dirty" indicator.

**Comment by JohnRDOrazio** [comment]
I agree that the "dirty state" indicator should not necessarily have to show that there is an undo stack, it should simply indicate the actual dirty state of the editor. If the file contents are the same, it should not show as dirty.

If changes have been made which result in the exact same state as that which is saved, editor should not show as dirty.

If I then hit "CTRL-Z" or "undo", and this re-introduces a change such that the current state no longer coincides with the saved state, then the indicator should show dirty again. If I then hit "CTRL-Y" or "redo" and this brings the state of the file back in line with the saved state, then the dirty indicator should go away again. It should not indicate an undo stack, it should simply indicate the actual dirty state of the file in the editor.

**Comment by clink-aaron** [comment]
Coming from vim, this is an extremely annoying bug with vscode. Is there any update on fixing this issue?

**Comment by GetPsyched** [comment]
> Nuts that this is still a problem after 4 years.

@runfaj, make that almost 8 years!

This should AT LEAST be a configuration option.

I hate it when devs mention usability/user experience arguments before anyone has even mentioned anything about making it the default behaviour.

**Comment by LordXerus** [comment]
@bpasero How large can the undo stack get? Does it get shrunken?

I think we can take a few approaches to reduce the amount of (expensive) full file comparisons.

First, size or line number check.

Then, consider that if the file is dirty, then at least one line is dirty. You could try something like this

On dirty:

1. Store line number to dirty line
2. Store original line
3. Store pointer to entry on undo stack

On edit:

1. Size/Line number check
2. Check if dirty line was changed
3. Compare to original line
4. While clean, replay line changes from undo stack (compare file?)
5. Then maybe run the line through the `On Dirty` procedure

Should work right?
